src/DataStructures/Tensor/Expressions/Contract.hpp

```diff
@@ -172,13 +176,296 @@ struct TensorContract
+  /// components \f${L^{a}}_{acb}\f$ for all values of \f$a\f$. This function
+  /// first constructs the list of generic indices (TensorIndexs) of the
+  /// uncontracted LHS, then uses a series of helper functions to compute a
+  /// mapping from (1) the storage indices of the components in the contracted
+  /// LHS tensor to (2) their corresponding lists of storage indices of
+  /// components in the uncontracted LHS tensor that need to be summed to
+  /// compute each contracted LHS component. Finally, the `ComputeContraction`
+  /// helper struct is used to compute the contracted component at
+  /// `lhs_storage_index` by leveraging this precomputed map's lists of indices
+  /// to sum for each contracted LHS component's storage index.
+  ///
+  /// \tparam LhsStructure the Structure of the contracted LHS tensor
+  /// \tparam LhsIndices the TensorIndexs of the contracted LHS tensor, e.g.
+  /// `ti_a_t`, `ti_b_t`, `ti_c_t`
+  /// \param lhs_storage_index the storage index of the contracted LHS tensor
+  /// component to retrieve
+  /// \return the value of the component at `lhs_storage_index` in the
+  /// contracted LHS tensor
   template <typename LhsStructure, typename... LhsIndices>
   SPECTRE_ALWAYS_INLINE decltype(auto) get(
       const size_t lhs_storage_index) const {
-    const std::array<size_t, num_tensor_indices>& new_tensor_index =
-        LhsStructure::template get_canonical_tensor_index<num_tensor_indices>(
-            lhs_storage_index);
-    return get<LhsIndices...>(new_tensor_index);
+    constexpr size_t contracted_lhs_num_components = LhsStructure::size();
+    using uncontracted_lhs_tensorindex_list =
+        get_uncontracted_lhs_tensorindex_list<LhsIndices...>;
+    using uncontracted_lhs_structure =
+        typename LhsTensorSymmAndIndices<ArgsList,
+                                         uncontracted_lhs_tensorindex_list,
+                                         Symm, IndexList>::structure;
+
+    constexpr std::make_index_sequence<contracted_lhs_num_components> map_seq{};
+
+    // A map from contracted LHS storage indices to lists of uncontracted LHS
+    // storage indices of components to sum for contraction
+    constexpr std::array<std::array<size_t, first_contracted_index::dim>,
+                         contracted_lhs_num_components>
+        map_of_components_to_sum =
+            get_map_of_components_to_sum<contracted_lhs_num_components,
+                                         uncontracted_lhs_structure,
+                                         LhsStructure>(map_seq);
+
+    // This returns the value of the component stored at `lhs_storage_index` in
+    // the contracted LHS tensor
+    return ComputeContraction<uncontracted_lhs_structure,
+                              uncontracted_lhs_tensorindex_list,
+                              contracted_lhs_num_components, decltype(t_),
+                              0>::apply(map_of_components_to_sum, t_,
```

Comment:
moving the template parameters to `apply` will mean you might need to do `::template apply<...>(...)`